kotlinx-coroutines-core/common/src/flow/terminal/Reduce.kt

```diff
@@ -47,33 +48,43 @@ public suspend inline fun <T, R> Flow<T>.fold(
 }
 
 /**
- * The terminal operator, that awaits for one and only one value to be published.
+ * The terminal operator that awaits for one and only one value to be emitted.
  * Throws [NoSuchElementException] for empty flow and [IllegalStateException] for flow
  * that contains more than one element.
  */
 public suspend fun <T> Flow<T>.single(): T {
     var result: Any? = NULL
     collect { value ->
-        if (result !== NULL) error("Expected only one element")
+        require(result == NULL) { "Flow has more than one element" }
         result = value
     }
 
-    if (result === NULL) throw NoSuchElementException("Expected at least one element")
-    @Suppress("UNCHECKED_CAST")
+    if (result === NULL) throw NoSuchElementException("Flow is empty")
     return result as T
 }
 
 /**
- * The terminal operator, that awaits for one and only one value to be published.
- * Throws [IllegalStateException] for flow that contains more than one element.
+ * The terminal operator that awaits for one and only one value to be emitted.
+ * Returns the single value or `null`, if the flow was empty or emitted more than one value.
  */
-public suspend fun <T: Any> Flow<T>.singleOrNull(): T? {
-    var result: T? = null
+public suspend fun <T> Flow<T>.singleOrNull(): T? {
+    var result: Any? = NULL
     collect { value ->
-        if (result != null) error("Expected only one element")
-        result = value
+        /*
```

Comment:
The `singleOrNull` operator shall be written via `collectWhile` to terminate yearly with a `null` result when the second value is encountered. This way, it will be consistent with stdlib `Sequence.singleOrNull` and with `Flow.single()` that also terminates early on the second item.